merkle/src/merkle.rs

```diff
@@ -586,72 +623,216 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
+                active_store.push(last_node);
 
                 width += 1;
-                j += 1;
             }
 
-            // elements are in [i..j] and they are even
-            let layer: Vec<_> = if j == self.leaves.len() {
-                self.leaves
-                    .read_range(i..j)
-                    .par_chunks(2)
-                    .map(|v| {
-                        let lhs = v[0].to_owned();
-                        let rhs = v[1].to_owned();
-                        A::default().node(lhs, rhs, height)
-                    })
-                    .collect()
-            } else if i >= self.leaves.len() {
-                self.top_half
-                    .read_range(i - self.leaves.len()..j - self.leaves.len())
+            // We read the `width` nodes of the current `level` from `read_store` and
+            // write (half of it) in the `write_store` (which contains the next level).
+            // Both `read_start` and `write_start` are "local" indexes with respect to
+            // the `read_store` and `write_store` they are accessing.
+            let (read_store_lock, write_store_lock, read_start, write_start) = if level == 0 {
+                // The first level is in the `leaves`, which is all it contains so the
+                // next level to write to will be in the `top_half`. Since we are "jumping"
+                // from one `Store` to the other both read/write start indexes start at zero.
+                (leaves_lock.clone(), top_half_lock.clone(), 0, 0)
+            } else {
+                // For all other levels we'll read/write from/to the `top_half` adjusting the
+                // "global" index to access this `Store` (offsetting `leaves` length). All levels
+                // are contiguous so we read/write `width` nodes apart.
+                let read_start = level_node_index - leaves_lock.read().unwrap().len();
+                (
+                    top_half_lock.clone(),
+                    top_half_lock.clone(),
+                    read_start,
+                    read_start + width,
+                )
+            };
+            // FIXME: Maybe just remove `write_store_lock` and always access `top_half_lock`
+            // directly if it makes it more readable.
+
+            // Allocate `width` indexes during operation (which is a negligible memory bloat
+            // compared to the 32-bytes size of the nodes stored in the `Store`s) and hash each
+            // pair of nodes to write them to the next level in concurrent threads.
+            // Process `chunk_size` nodes in each thread at a time to reduce contention, optimized
+            // for big sector sizes (small ones will just have one thread doing all the work).
+            let chunk_size = 1024;
+            debug_assert_eq!(chunk_size % 2, 0);
+            Vec::from_iter((read_start..read_start + width).step_by(chunk_size))
+                .par_chunks(1)
```

Comment:
This should be `par_iter` I believe

Reply:
yes, now that the chunk size is 1 it makes sense to just iterate